src/D2L.CodeStyle.Analyzers/Async/Generator/SyncGenerator.cs

```diff
@@ -0,0 +1,182 @@
+			.SelectMany( al => al.Attributes );
+
+		foreach( var attribute in attributes ) {
+			// We have the semantic model now, so make sure this method has
+			// the right attribute.
+			if( IsGenerateSyncAttribute( context.SemanticModel, attribute, cancellationToken ) ) {
+				return methodDeclaration;
+			}
+		}
+
+		return null;
+	}
+
+	internal static bool IsGenerateSyncAttribute(
+		SemanticModel model,
+		AttributeSyntax attribute,
+		CancellationToken cancellationToken
+	) {
+		var attributeConstructorSymbol = model.GetSymbolInfo( attribute, cancellationToken ).Symbol as IMethodSymbol;
+
+		if( attributeConstructorSymbol == null ) {
+			return false;
+		}
+
+		return attributeConstructorSymbol.ContainingType.ToDisplayString()
+			== "D2L.CodeStyle.Annotations.GenerateSyncAttribute";
+	}
+
+	private static MethodGenerationResult? GenerateSyncMethod(
+		(MethodDeclarationSyntax MethodDeclaration, Compilation Compilation) data,
+		CancellationToken cancellationToken
+	) {
+		if( !data.Compilation.ContainsSyntaxTree( data.MethodDeclaration.SyntaxTree ) ) {
+			return null;
+		}
+
+		var model = data.Compilation.GetSemanticModel( data.MethodDeclaration.SyntaxTree );
+		var methodSymbol = model.GetDeclaredSymbol( data.MethodDeclaration, cancellationToken );
+
+		if( methodSymbol == null || methodSymbol.Kind == SymbolKind.ErrorType ) {
+			return null;
+		}
+
+		var transformer = new AsyncToSyncMethodTransformer( model, cancellationToken );
+
+		var transformResult = transformer.Transform( data.MethodDeclaration );
+
+		return new(
+			Original: data.MethodDeclaration,
+			GeneratedSyntax: transformResult.Success ? transformResult.Value!.ToFullString() : "",
+			Diagnostics: transformResult.Diagnostics
+		);
+	}
+
+	private static IEnumerable<FileGenerationResult> GenerateFiles(
+		ImmutableArray<MethodGenerationResult> methodResults,
+		CancellationToken cancellationToken
+	) {
+		var methodsByFile = methodResults
+			.GroupBy( static mr => mr.Original.SyntaxTree.GetCompilationUnitRoot() )
+			.OrderBy( static mr => mr.First().Original.SyntaxTree.FilePath )
+			.ToImmutableArray();
+
+		var hintNames = SourceFilePathHintNameGenerator.Create(
+			// We're assuming every root has a tree here -- when is that false?
+			// Generated code?
+			methodsByFile.Select( static grp => grp.Key.SyntaxTree.FilePath ).ToImmutableArray()
+		).GetHintNames();
+
+		var files = methodsByFile.Zip(
+			hintNames,
+			static ( fileData, name ) => ( HintName: name, Data: fileData )
+		);
+
+		foreach( var file in files ) {
+			if( cancellationToken.IsCancellationRequested ) {
+				yield break;
+			}
+
+			var generatedMethods = ImmutableArray.CreateBuilder<(TypeDeclarationSyntax, string)>();
+			var diagnostics = ImmutableArray.CreateBuilder<Diagnostic>();
+
+			foreach( var method in file.Data ) {
+				diagnostics.AddRange( method.Diagnostics );
+				generatedMethods.Add( ((method.Original.Parent as TypeDeclarationSyntax)!, method.GeneratedSyntax ) );

```

Comment:
Should this just be a cast?

```suggestion
				generatedMethods.Add( ((TypeDeclarationSyntax)method.Original.Parent, method.GeneratedSyntax ) );
```

Reply:
Not with nullable on apparently. What I put is more explicit about discarding the `?` I guess.